lib/network/internal.nix

```diff
@@ -0,0 +1,220 @@
+      _encode 3232235521
+      => "192.168.0.1"
+      ```
+
+      # Type
+
+      ```
+      _encode :: Int -> String
+      ```
+
+      # Arguments
+
+      - [num] A decoded integer representation of an IPv4 address.
+
+      # Throws
+
+      - If the argument is less than zero.
+      - If the argument is greater than or equal to 2^32.
+    */
+    _encode =
+      num:
+      if num < 0 then
+        throw "lib.network.ipv4._encode: ${toString num} cannot be encoded into an IPv4 address."
+      else if num > 4294967295 then
+        throw "lib.network.ipv4._encode: ${toString num} is too large to encode into an IPv4 address."
+      else
+        concatStringsSep "." (
+          map (x: toString (mod (num / x) 256)) (reverseList (genList (x: common.pow 2 (x * 8)) 4))
+        );
+
+    /**
+      Extracts a prefix length from a CIDR and verifies it is valid.
+
+      If an IP address is given without a CIDR, then a prefix length of 32 is returned.
+
+      # Example
+
+      ```nix
+      _verifyPrefixLength "192.168.0.1/24"
+      => 24
+      _verifyPrefixLength "192.168.0.1"
+      => 32
+      ```
+
+      # Type
+
+      ```
+      _verifyPrefixLength :: String => String
+      ```
+
+      # Arguments
+
+      - [cidr] An IPv4 CIDR.
+
+      # Throws
+
+      - If a CIDR was given with a slash but no prefix length following.
+      - If there were multiple slashes in the CIDR.
+    */
+    _verifyPrefixLength =
+      cidr:
+      let
+        splitCidr = lib.splitString "/" cidr;
+      in
+      if (builtins.length splitCidr) == 1 then
+        "32"
+      else if (builtins.length splitCidr) > 2 then
+        throw "lib.network.ipv4: Could not verify prefix length for CIDR ${cidr}."
+      else
+        let
+          afterSlash = elemAt splitCidr 1;
+        in
+        if afterSlash == "" then
+          throw "lib.network.ipv4: CIDR ${cidr} has no prefix length."
+        else if toInt afterSlash > 32 || toInt afterSlash < 0 then
+          throw "lib.network.ipv4: CIDR ${cidr} has an out of bounds prefix length, ${afterSlash}."
+        else
+          afterSlash;
+
+    _verifyAddress =
+      cidr:
+      let
+        splitCidr = lib.splitString "/" cidr;
+        address = elemAt splitCidr 0;
+        splitAddress = lib.splitString "." address;
+        intOctets = map toInt splitAddress;
+      in
+      if (builtins.length splitAddress) != 4 then
+        throw "lib.network.ipv4: CIDR ${cidr} is not of the correct form."
+      else if any (x: x == "") splitAddress then
+        throw "lib.network.ipv4: CIDR ${cidr} has an empty octet."
+      else if !(all (x: x >= 0 && x <= 255)) intOctets then
+        throw "lib.network.ipv4: CIDR ${cidr} has an out of bounds octet."
+      else
+        address;
+
+    /**
+      Given an IP address and prefix length, creates an attribute set of network parameters.
+
+      # Example
+
+      ```nix
+      _parse "192.168.0.1/24"
+      => {
+        address = "192.168.0.1";
```

Comment:
Shouldn't this return something like `[ 192 168 0 1 ]`, such that other functions can more easily work with the data instead of having to re-parse it?

Reply:
After thinking about it, that is a good idea. But I think this should be stored in an internal variable like `_address` with the string representation stored in `address`. The reason being that the general use case for the IP address is in it's string representation. But yeah having the internal representation as a list of ints is probably the best state to store it internally.